cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go

```diff
@@ -45,12 +45,9 @@ import (
 const DefaultModule = "main"
 
 type bicepDeploymentDetails struct {
-	// Template is the template to deploy during the deployment operation.
-	Template azure.RawArmTemplate
+	CompiledBicep *compileBicepResult
```

Comment:
`compileBicepResult` also contains a `Parameters`.  I'm not sure how we rationalize the Parameters filed in that type with the one in `bicepDeploymentDetails`.  Could we undo this part of the change and have this structure continue to have the shape as before?  Failing that, I think we should remove `Parameters` from here and then ensure that when we build a `bicepDeploymentDetails` instance, we always have a non nil `Parameters`.

Reply:
CompiledBicepResults contains Parameters but only when a `.bicepparams` file was compiled, otherwise that field is nil.

azd doesn't know (and shouldn't care) about what type of bicep was compiled until the actual compilation. 
After that, azd would take the parameters from the compilation (from bicepparams) or, otherwise, it will get the the parameters.json and produce the `final-parameters`.  Those final-parameters are the `bicepDeploymentDetails.Parameters`

Would it help if we rename the Parameters within `CompiledBicepResults` to `CompiledTimeParameters` ?

The reason why I made this change is because I need to combine the parameters-definitions and the final-parameters to produce the `resultParameters` that are returned from a previous deployment.   There are 3 ways of doing this:
 1. Making `bicep.plan()` to include the entire compileResult when it returns. This is what I did. This way the template is already parsed into parameter definitions and ready to use.
 2. Making `bicep.plan()` to include the template next to the `azure.RawArmTemplate`.  The rawArmTemplate is used for the deployment and the ArmTemplate is used to inspect the parameter definitions. I didn't like this one because there would be 2 templates in `bicepDeploymentDetails `
 3. Do not update the bicep.plan() return and use/parse `azure.RawArmTemplate` back to  ArmTemplate to check the parameter definitions.  I didn't like this because of the extra unmarshalling process to create the Type from the serialized data, specially because that Type was previously produced during plan.

Reply:
> The reason why I made this change is because I need to combine the parameters-definitions and the final-parameters to produce the resultParameters that are returned from a previous deployment.

I was trying to convince myself of this and started to apply a version of this change locally to better understand the problem you described here.  As I worked through the problem, I realized I don't fully understand how we end up doing the parameter comparison to determine if the deployment we intend to do matches the deployment that was already done.  It looks like in the diff here we are concerned about ensuring defaults from the template, among other things, and it's not 100% clear to me why that is the case.  Is this because the parameters dictionary in the completed deployment object includes the values that came from defaults, but that are not in `bicepDeploymentData.Parameters`?

Looking at a completed deployment, one problem I foresee is that the deployment object doesn't retain the values for secure parameters.  I think given your implementation, we'll treat this as a cache miss and just do the deployment again, which is the right thing to do given we can't know but does mean this feature doesn't work with `@secure` parameters.

Part of me wonders if we should be doing something like computing a "parameter hash" (based on the value of `bicepDeploymentData.Parameters`, if we can come up with some canonicalization strategy, say the SHA256 hash of the JSON or something and applying that as a tag (`azd-parameter-hash`) on the deployment object, and when this is present just comparing these hashes as we would template hashes?  If we went down that road, since the data is public, we might salt the message we hash so that you can't compare between two deployments and tell "yes these had the same parameters".

If you help me understand what the strategy, we are using to compare the properties, I can see if there is a way for us to implement it cleanly without the above refactoring.

Something else I realized - we should probably only do this skip when the previous deployment is "successful".  If it failed, the user may want to submit the same deployment again and have it run (say, for example, they enabled some RP or waited for an outage to complete).

Reply:
- removed parameters field from `bicepDeploymentDetails` and make sure that bicepDeploymentDetails.CompiledBicep.Parameters is where we keep the parameters
- Updated strategy for comparing parameters using hash